### ISSUE TITLE
Rule request: no-implicit-null-types

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [Microsoft/tslint-microsoft-contrib#358]
*For teams with a large code base, turning on `--strictNullChecks` can take a while. Having rules that do part of the work would help incrementally enabling it.*

If a variable or parameter is assigned null or undefined, its type should reflect that.

Bad:

```typescript
class Person {
    private name;
    public setName(name: string | undefined) {
        this.name = name; // This should be an error
    }
}
```

**Comment by JoshuaKGoldberg** [comment]
This is grander in scale than what a linter can do well. Addressing this would need access to the TypeScript language service.

I've started work on this in [TypeStat](https://github.com/JoshuaKGoldberg/TypeStat) instead.
